sonar-kotlin-gradle/src/main/java/org/sonarsource/kotlin/gradle/checks/DependencyGroupedCheck.kt

```diff
@@ -0,0 +1,143 @@
+        "rhinoPluginRhinoClasspath",
+        "runtime",
+        "runtimeClasspath",
+        "runtimeElements",
+        "runtimeOnly",
+        "signatures",
+        "testAnnotationProcessor",
+        "testCompile",
+        "testCompileClasspath",
+        "testCompileOnly",
+        "testImplementation",
+        "testRuntime",
+        "testRuntimeClasspath",
+        "testRuntimeOnly",
+        "zinc",
+    )
+}
+
+@Rule(key = "S6629")
+class DependencyGroupedCheck : AbstractCheck() {
+
+    override fun visitCallExpression(expression: KtCallExpression, kotlinFileContext: KotlinFileContext) {
+        if (getFunctionNameOrNull(expression) != REF_NAME_DEPENDENCIES) return
+        checkDependencyHandlerScopeLambda(expression, kotlinFileContext)
+    }
+
+    private fun checkDependencyHandlerScopeLambda(expression: KtCallExpression, kotlinFileContext: KotlinFileContext) {
+        getLambdaBlock(expression)?.acceptChildren(DependencyHandlerScopeLambdaCheck(kotlinFileContext))
+    }
+
+    private inner class DependencyHandlerScopeLambdaCheck(
+        private val kotlinFileContext: KotlinFileContext
+    ) : KtVisitorVoid() {
+
+        private var lastNameInOrder: String? = null
+
+        private val visitedNames = mutableSetOf<String>()
+
+        override fun visitCallExpression(expression: KtCallExpression) {
+            val functionReference = expression.getCalleeExpressionIfAny() as? KtNameReferenceExpression ?: return
+            val functionName = functionReference.getReferencedName()
+            if (!REF_NAME_DEPENDENCY_HANDLER_SCOPE_EXTENSIONS.contains(functionName)) return
+
+            if (functionName == REF_NAME_CONSTRAINTS) {
+                checkDependencyHandlerScopeLambda(expression, kotlinFileContext)
+            } else {
+                checkFunctionReferenceOrder(functionReference)
+            }
+        }
+
+        fun checkFunctionReferenceOrder(functionReference: KtNameReferenceExpression) {
+            val functionName = functionReference.getReferencedName()
+            if (lastNameInOrder != null && functionName != lastNameInOrder && visitedNames.contains(functionName)) {
+                kotlinFileContext.reportIssue(
+                    functionReference,
+                    message {
+                        +"Group "
+                        code(functionName)
+                        +" dependencies"
+                    }
+                )
+            } else {
+                lastNameInOrder = functionName
+            }
+            visitedNames.add(functionName)
+        }
+    }
+}
+
+private fun getFunctionNameOrNull(expression: KtCallExpression) =
+    (expression.getCalleeExpressionIfAny() as? KtNameReferenceExpression)?.getReferencedName()
+
+private fun getLambdaBlock(expression: KtCallExpression): KtBlockExpression? {
+    val lambdaArg = expression.valueArguments.lastOrNull() as? KtLambdaArgument
+    return lambdaArg?.getLambdaExpression()?.bodyExpression
```

Comment:
Don't have strong opinion but you can have a single expression here to look more "functional"

Reply:
I'd like to leave it so just to break down the complexity of the expression. I think that's better readable.